src/common/download.h

```diff
@@ -32,5 +32,13 @@
 
 namespace tools
 {
-  bool download(const std::string &path, const std::string &url);
+  struct download_thread_control;
+  typedef std::shared_ptr<download_thread_control> download_async_handle;
+
+  bool download(const std::string &path, const std::string &url, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress = NULL);
+  download_async_handle download_async(const std::string &path, const std::string &url, std::function<void(const std::string&, const std::string&, bool)> result, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress = NULL);
+  bool download_error(download_async_handle h);
```

Comment:
Also, a `const download_async_handle&` signature prevents an atomic ref count when calling the function.